Bottom: 6aa1368ff54e4ad9fc7fb09ac5fc38401269cb74
Top:    044087992f81f64cfce689020cb91439d2b39388
Author: Peter Clifton <peter@clifton-electronics.co.uk>
Date:   2014-07-10 19:01:30 +0100

polygon.c: Add option to include drills holes (from vias / pins) in the board outline


---

diff --git a/src/hid/gtk/gtkhid-gl.c b/src/hid/gtk/gtkhid-gl.c
index 558c16a..d2834bd 100644
--- a/src/hid/gtk/gtkhid-gl.c
+++ b/src/hid/gtk/gtkhid-gl.c
@@ -1569,7 +1569,7 @@ ensure_board_outline (void)
     if (PCB->Data->outline != NULL)
       poly_Free (&PCB->Data->outline);
 
-    PCB->Data->outline = board_outline_poly ();
+    PCB->Data->outline = board_outline_poly (false);
     PCB->Data->outline_valid = true;
   }
 }
diff --git a/src/polygon.c b/src/polygon.c
index e91f990..ce68a11 100644
--- a/src/polygon.c
+++ b/src/polygon.c
@@ -2013,6 +2013,22 @@ line_outline_callback (const BoxType * b, void *cl)
   return 1;
 }
 
+static int
+pv_outline_callback (const BoxType * b, void *cl)
+{
+  PinType *pv = (PinType *)b;
+  struct clip_outline_info *info = cl;
+  POLYAREA *np, *res;
+
+  if (!(np = CirclePoly (pv->X, pv->Y, pv->DrillingHole / 2)))
+    return 0;
+
+  poly_Boolean_free (info->poly, np, &res, PBO_SUB);
+  info->poly = res;
+
+  return 1;
+}
+
 static void
 delete_piece_cb (gpointer data, gpointer userdata)
 {
@@ -2034,7 +2050,7 @@ delete_piece_cb (gpointer data, gpointer userdata)
   poly_Free (&piece);
 }
 
-POLYAREA *board_outline_poly (void)
+POLYAREA *board_outline_poly (bool include_holes)
 {
   int i;
   int count;
@@ -2103,6 +2119,12 @@ POLYAREA *board_outline_poly (void)
   r_search (Layer->line_tree, &region, NULL, line_outline_callback, &info);
   r_search (Layer->arc_tree,  &region, NULL, arc_outline_callback, &info);
 
+  if (include_holes)
+    {
+      r_search (PCB->Data->pin_tree, &region, NULL, pv_outline_callback, &info);
+      r_search (PCB->Data->via_tree, &region, NULL, pv_outline_callback, &info);
+    }
+
   clipped = info.poly;
 
   /* Now we just need to work out which pieces of polygon are inside
diff --git a/src/polygon.h b/src/polygon.h
index 30e5ea3..a02fd99 100644
--- a/src/polygon.h
+++ b/src/polygon.h
@@ -88,5 +88,5 @@ bool MorphPolygon (LayerType *, PolygonType *);
 void NoHolesPolygonDicer (PolygonType *p, const BoxType *clip,
                           void (*emit) (PLINE *, void *), void *user_data);
 void PolyToPolygonsOnLayer (DataType *, LayerType *, POLYAREA *, FlagType);
-POLYAREA *board_outline_poly (void);
+POLYAREA *board_outline_poly (bool include_holes);
 #endif
